### PR TITLE
More avoidance for loadAuthorizedIndices

### PR DESCRIPTION
**Pull request by ywangd**
This PR replaces the concrete list of authorizeIndices with a lazily load class
so that loading is not triggered if not necessary. 

For example, a search targeting a concret name, `GET index/_search` no longer
triggers the loading. But names with wildcard will still trigger the loading.

However if we organize the indices with data streams and alias, it is possible
to achieve similar wildcard effect while still avoid the loading. That is,
searches like `GET alias/_search` or `GET data_stream/_search` will *not*
trigger the loading even they target multiple indices behind the single name.

### CODE REVIEW

File: x-pack/plugin/security/src/main/java/org/elasticsearch/xpack/security/authz/AuthorizationService.java [hunk 1]
```diff
@@ -405,38 +405,28 @@ private void authorizeAction(
...
         } else if (isIndexAction(action)) {
             final Metadata metadata = clusterService.state().metadata();
-            final AsyncSupplier<Set<String>> authorizedIndicesSupplier = new CachingAsyncSupplier<>(authzIndicesListener -> {
-                LoadAuthorizedIndiciesTimeChecker timeChecker = LoadAuthorizedIndiciesTimeChecker.start(requestInfo, authzInfo);
-                authzEngine.loadAuthorizedIndices(
-                    requestInfo,
-                    authzInfo,
-                    metadata.getIndicesLookup(),
-                    authzIndicesListener.map(authzIndices -> {
-                        timeChecker.done(authzIndices);
```
Comment (ywangd) [inline x-pack/plugin/security/src/main/java/org/elasticsearch/xpack/security/authz/AuthorizationService.java:415]: Logging for slow loadAuthorizedIndices is not easily relocatable with the proposed changes because the logic is now in server package and does not know anything about user or roles. One possibility is to pass a string description to that class just for logging purpose.

**Comment by ywangd** [comment]
@elasticmachine update branch

File: x-pack/plugin/core/src/main/java/org/elasticsearch/xpack/core/security/authz/permission/IndicesPermission.java [hunk 2]
```diff
@@ -159,11 +159,15 @@ public boolean hasFieldOrDocumentLevelSecurity() {
...
-                && bwcSpecialCaseMatcher.test(indexAbstraction.getName()));
+        if (grantMappingUpdatesOnIndices.isEmpty() && grantMappingUpdatesOnRestrictedIndices.isEmpty()) {
+            return indexAbstraction -> nameMatcher.test(indexAbstraction.getName());
+        } else {
+            final StringMatcher bwcSpecialCaseMatcher = indexMatcher(grantMappingUpdatesOnIndices, grantMappingUpdatesOnRestrictedIndices);
+            return indexAbstraction -> nameMatcher.test(indexAbstraction.getName())
+                || (indexAbstraction.getType() != IndexAbstraction.Type.DATA_STREAM
+                    && (indexAbstraction.getParentDataStream() == null)
+                    && bwcSpecialCaseMatcher.test(indexAbstraction.getName()));
+        }
```
Comment (ywangd) [inline x-pack/plugin/core/src/main/java/org/elasticsearch/xpack/core/security/authz/permission/IndicesPermission.java:170]: Just realised that this block of changes is not related to the purpose of this PR. It's a minor optimisation for `allowedIndicesMatcher`

File: x-pack/plugin/core/src/main/java/org/elasticsearch/xpack/core/security/authz/AuthorizationEngine.java [hunk 3]
```diff
@@ -152,13 +154,20 @@ void authorizeIndexAction(
...
         AuthorizationInfo authorizationInfo,
         Map<String, IndexAbstraction> indicesLookup,
         ActionListener<Set<String>> listener
     );
 
+    Predicate<IndexAbstraction> predicateForAuthorizedIndices(
+        RequestInfo requestInfo,
+        AuthorizationInfo authorizationInfo,
+        Metadata metadata
+    );
```
Comment (albertzaharovits) [inline x-pack/plugin/core/src/main/java/org/elasticsearch/xpack/core/security/authz/AuthorizationEngine.java:169]: Do you see an inherent advantage with using a predicate rather than a set?
(besides the delayed evalution, which I think should work for the set variant)

File: x-pack/plugin/core/src/main/java/org/elasticsearch/xpack/core/security/authz/permission/IndicesPermission.java (hunk 2 above)
Comment (albertzaharovits) [reply x-pack/plugin/core/src/main/java/org/elasticsearch/xpack/core/security/authz/permission/IndicesPermission.java:170]: I think these sort of small things are a good thing, I don't think they harm the readability.

File: x-pack/plugin/security/src/main/java/org/elasticsearch/xpack/security/authz/RBACEngine.java [hunk 4]
```diff
@@ -692,6 +704,29 @@ GetUserPrivilegesResponse buildUserPrivilegesResponseObject(Role userRole) {
...
+                if (predicate.test(indexAbstraction)) {
+                    return true;
+                } else {
+                    return indexAbstraction.getParentDataStream() != null && predicate.test(indexAbstraction.getParentDataStream());
+                }
+            } else {
+                return indexAbstraction.getType() != IndexAbstraction.Type.DATA_STREAM && predicate.test(indexAbstraction);
+            }
+        };
+    }
```
Comment (albertzaharovits) [inline x-pack/plugin/security/src/main/java/org/elasticsearch/xpack/security/authz/RBACEngine.java:728]: This predicate when evaluated over all the names in the cluster state will end-up being evaluated twice over data stream names, no?
But maybe it's not a problem?

**Comment by albertzaharovits** [comment]
I think delaying evaluating the `authorizedIndices` will turn out to be a great optimization in practice because many requests do not contain wildcards.
I'm not sure about the switch to a predicate though (https://github.com/elastic/elasticsearch/pull/81237#discussion_r778974797) . I need to think it through a bit more, but maybe you can clarify your thoughts about this particular change.

**Comment by albertzaharovits** [comment]
TBC I'm generally in favor of this change.

File: x-pack/plugin/core/src/main/java/org/elasticsearch/xpack/core/security/authz/permission/IndicesPermission.java (hunk 2 above)
Comment (ywangd) [reply x-pack/plugin/core/src/main/java/org/elasticsearch/xpack/core/security/authz/permission/IndicesPermission.java]: It's not about readability. It is a very minor optimisation to not create `bwcSpecialCaseMatcher` and subsequently using it if we know it's not necessary. That said, the change has nothing to do with this PR, it's now reverted.

File: x-pack/plugin/core/src/main/java/org/elasticsearch/xpack/core/security/authz/AuthorizationEngine.java (hunk 3 above)
Comment (ywangd) [reply x-pack/plugin/core/src/main/java/org/elasticsearch/xpack/core/security/authz/AuthorizationEngine.java:169]: Yes because it avoids building the Set for `Set#contains` calls, e.g.:

https://github.com/elastic/elasticsearch/blob/861dd90c8719742f004a882f09415d8fbab1fc12/server/src/main/java/org/elasticsearch/cluster/metadata/IndexAbstractionResolver.java#L85

https://github.com/elastic/elasticsearch/blob/861dd90c8719742f004a882f09415d8fbab1fc12/server/src/main/java/org/elasticsearch/cluster/metadata/IndexAbstractionResolver.java#L138

These are the places where the requested indices do not contain any wildcard. In this case, building the Set (of authorized indices) is not necessary. Testing directly against the predicate is most likely more efficient.

The predicate does not help when we do need iterate through the Set, e.g.:

https://github.com/elastic/elasticsearch/blob/861dd90c8719742f004a882f09415d8fbab1fc12/server/src/main/java/org/elasticsearch/cluster/metadata/IndexAbstractionResolver.java#L110

In this case, we have no choice but build the Set from the predicate. This step has the "double evaluation" issue over data stream names as you raised in the following comment. I'll address it over there.

File: x-pack/plugin/security/src/main/java/org/elasticsearch/xpack/security/authz/RBACEngine.java (hunk 4 above)
Comment (ywangd) [reply x-pack/plugin/security/src/main/java/org/elasticsearch/xpack/security/authz/RBACEngine.java:728]: You are right. I didn't think of it. When privilege is _not_ granted for backing index names (the most probable case), running a backing index name through the predicate ends up always evaluating both the index name and the data stream name.

I don't know how bad it is. But definitely _not_ ideal. It is a bit trickly to solve. I came up with following solution:

1. Reversed the order of testing of index name and data stream name. It now always check data stream name first. This should be better for testing concrete requested index names because (1) The requested indices are most likely data streams instead of backing indices; (2) Privileges are likely granted over data streams. Therefore, when the concrete index name should be authorized, it will likely be a single test with data stream name or a single test with the concrete index name (if it does not belong to a data stream). When it should not be authorized, both tests can run. But this will be a 403 error case, performance is less of an issue here. 

2. NOTE the 1st point does not help when we must evaluated through all indices, i.e. when wildcard must be expanded. In this case, the caller of the predicate nows does its own optimisation to add all backing indices of a data stream if the data stream is authorized. This avoids running the predicate for backing index names.

**Comment by ywangd** [comment]
@albertzaharovits I made a rather big change based on the discussion. It is still along the agreement we reached last time and overall a simplication. Instead of changing the signatures of existing Classes (`AuthorizationEngine`, `IndexAbstractionResolver` etc), the PR now introduces a new `Set` implemetation which lazily builds the authorized indices when necessary, but otherwise uses predicate for `contains` test. So the idea does not change, but the changeset is much smaller. Here is a summary for how it should work for different scenarios:
1. Concrete names, e.g. `index1`, or `index1,index2,index3`. Performance should be much better because the predicate testing for the individual names should much faster than loading all authorized indices. [1]
2. Simple wildcard, e.g. `*`, or `logs-*,metrics-*`. No change to performance because the authorizedIndices set will just be built when there are genuine wildcard expansion.
3. The more interesting case is a mix of concrete names and wildcards, e.g. `index1,logs-*`. The performance could be slightly worse because the code first runs the predicate for `index1` and then still needs to load `authorizedIndices` for `logs-*`. So the performance lost is one additional predicate execution. Obviously, the performance can keep degrade with longer list of concrete index names. [2]

I think the change should have a postive impact on performance overall because above Scenario 1 is very common and the potential gain is significant (especially when the cluster is large in size). Scenario 3 does have certain performance penalty but it is mitigated by (1) We can advise users to list wildcard patterns first in the requested names; (2) It is unlikely for the list of concrete names to be very long which means the performance drop should not be anyway significant (compared to loading authorizedIndices).

[1] An edge case of this scenario is that all requested names are for backing indices and the user does _not_ have privileges over data stream. In this case, the index pattern predicate can run twice for each give name (first for the data stream name, a second time for the backing index name). However, it still unlikely an issue because (1) it still saves time in loading authorized indices which is likely much more expensive than running the predicate couple more times; (2) We do not recommend granting privileges only over the backing indices. If the user does _not_ have privileges over the backing indices either, the request will end up be a 403 failure and performance is not critical in the failure case.

[2] We can avoid the additional predicate executions by looping through the requested names earlier and find out whether it has any wildcard. However, this loop itself also has its cost. And if the names are all concrete, it ends up doing useless work. In short, it is still a tradeoff between different scenarios.

File: x-pack/plugin/security/src/main/java/org/elasticsearch/xpack/security/authz/RBACEngine.java [hunk 5]
```diff
@@ -659,34 +670,65 @@ GetUserPrivilegesResponse buildUserPrivilegesResponseObject(Role userRole) {
...
+                    if (indexAbstraction.getType() != IndexAbstraction.Type.DATA_STREAM && predicate.test(indexAbstraction)) {
+                        indicesAndAliases.add(indexAbstraction.getName());
+                    }
+                }
             }
-        } else {
-            for (IndexAbstraction indexAbstraction : lookup.values()) {
-                if (indexAbstraction.getType() != IndexAbstraction.Type.DATA_STREAM && predicate.test(indexAbstraction)) {
-                    indicesAndAliases.add(indexAbstraction.getName());
+            final Set<String> authorizedIndices = Collections.unmodifiableSet(indicesAndAliases);
```
Comment (albertzaharovits) [inline x-pack/plugin/security/src/main/java/org/elasticsearch/xpack/security/authz/RBACEngine.java:711]: The `AuthorizedIndicesSet` should protect against modification to this, so maybe avoid the wrapping in an unmodifiable set.

File: x-pack/plugin/security/src/main/java/org/elasticsearch/xpack/security/authz/RBACEngine.java [hunk 6]
```diff
@@ -813,4 +855,96 @@ private static boolean isAsyncRelatedAction(String action) {
...
+        }
+
+        @Override
+        public boolean remove(Object o) {
+            throw new UnsupportedOperationException();
+        }
+
+        @Override
+        public boolean containsAll(Collection<?> c) {
+            return false;
```
Comment (albertzaharovits) [inline x-pack/plugin/security/src/main/java/org/elasticsearch/xpack/security/authz/RBACEngine.java:927]: I think we should implement this in some way.

File: x-pack/plugin/security/src/main/java/org/elasticsearch/xpack/security/authz/RBACEngine.java [hunk 7]
```diff
@@ -813,4 +855,96 @@ private static boolean isAsyncRelatedAction(String action) {
...
+        AuthorizedIndicesSet(Supplier<Set<String>> supplier, Predicate<String> predicate) {
+            this.supplier = Objects.requireNonNull(supplier);
+            this.predicate = Objects.requireNonNull(predicate);
+        }
+
+        private Set<String> getAuthorizedIndices() {
+            if (authorizedIndices == null) {
+                authorizedIndices = supplier.get();
+            }
+            return authorizedIndices;
```
Comment (albertzaharovits) [inline x-pack/plugin/security/src/main/java/org/elasticsearch/xpack/security/authz/RBACEngine.java:878]: I would **slightly** prefer we make this check thread safe, like in the `CachingAsyncSupplier`.
We're not using it in a multi-thread fashion, and I don't think we would ever do, and even if we did we only risk performance problems with building the set multiple times (but no correctness problems).
What do you think?

File: x-pack/plugin/security/src/main/java/org/elasticsearch/xpack/security/authz/AuthorizationService.java [hunk 8]
```diff
@@ -404,18 +407,11 @@ private void authorizeAction(
...
-                    requestInfo,
-                    authzInfo,
-                    metadata.getIndicesLookup(),
-                    authzIndicesListener.map(authzIndices -> {
-                        timeChecker.accept(authzIndices);
-                        return authzIndices;
-                    })
-                );
-            });
+            final AsyncSupplier<Set<String>> authorizedIndicesSupplier = new CachingAsyncSupplier<>(
```
Comment (albertzaharovits) [inline x-pack/plugin/security/src/main/java/org/elasticsearch/xpack/security/authz/AuthorizationService.java:410]: Do you think we still need the `CachingAsyncSupplier` wrapper magic?
I would ditch it, it's hard to follow and useless now.

**Review by albertzaharovits — APPROVED** [review]
LGTM this is looking cool!
Only had small points.

File: x-pack/plugin/security/src/main/java/org/elasticsearch/xpack/security/authz/RBACEngine.java (hunk 5 above)
Comment (ywangd) [reply x-pack/plugin/security/src/main/java/org/elasticsearch/xpack/security/authz/RBACEngine.java:711]: Yes it is not strictly necessary. The Set is passed into `timeChecker.accept` in the next line. I was pretending that if we don't know anything about `timeChecker` and want to have maximum protection against modification, then we need to wrap it in an `unmodifiableSet`. I think it was just me being unnecessarily paranoid. So I dropped the wrapping.

File: x-pack/plugin/security/src/main/java/org/elasticsearch/xpack/security/authz/RBACEngine.java (hunk 6 above)
Comment (ywangd) [reply x-pack/plugin/security/src/main/java/org/elasticsearch/xpack/security/authz/RBACEngine.java:927]: Thanks for catching it. I meant to but somehow missed it. Fixed by just delegating to `contains`

File: x-pack/plugin/security/src/main/java/org/elasticsearch/xpack/security/authz/RBACEngine.java (hunk 7 above)
Comment (ywangd) [reply x-pack/plugin/security/src/main/java/org/elasticsearch/xpack/security/authz/RBACEngine.java:878]: This Set is wrapped in a `CachingAsyncSupplier`, i.e. `resolvedIndicesAsyncSupplier` (not `authorizedIndicesSupplier` which I will remove, see below). The `resolvedIndicesAsyncSupplier` is what gets passed to other places so that this Set is never directly used by anything else. Since `resolvedIndicesAsyncSupplier` already guarantees thread safety, I think the Set itself can be kept simple. I added the following comment to the class to clarify its usage:
```
* NOTE that the lazy loading does not guarantee to run only once and is not meant to be used by multi-threads
* because loading multiple times can incur performance penalty (but not correctness).
```

File: x-pack/plugin/security/src/main/java/org/elasticsearch/xpack/security/authz/AuthorizationService.java (hunk 8 above)
Comment (ywangd) [reply x-pack/plugin/security/src/main/java/org/elasticsearch/xpack/security/authz/AuthorizationService.java:410]: I agree it is redundant now and should be removed.

File: x-pack/plugin/security/src/main/java/org/elasticsearch/xpack/security/authz/RBACEngine.java [hunk 9]
```diff
@@ -114,10 +117,16 @@
 
     private final CompositeRolesStore rolesStore;
     private final FieldPermissionsCache fieldPermissionsCache;
+    private final LoadAuthorizedIndicesTimeChecker.Factory authzIndicesTimerFactory;
 
-    public RBACEngine(Settings settings, CompositeRolesStore rolesStore) {
+    public RBACEngine(
+        Settings settings,
+        CompositeRolesStore rolesStore,
+        LoadAuthorizedIndicesTimeChecker.Factory authzIndicesTimerFactory
```
Comment (ywangd) [inline x-pack/plugin/security/src/main/java/org/elasticsearch/xpack/security/authz/RBACEngine.java:125]: Moving TimeChecker here means it is no longer applied to custom `AuthorizationEngine`. I think it is OK. The intention was always to measure the performance of RBACEngine.

File: docs/changelog/81237.yaml [hunk 10]
```diff
@@ -0,0 +1,5 @@
+pr: 81237
+summary: More avoidance for `loadAuthorizedIndices`
```
Comment (tvernum) [inline docs/changelog/81237.yaml:2]: Can we update this with a more user-friendly description?

File: x-pack/plugin/security/src/main/java/org/elasticsearch/xpack/security/authz/RBACEngine.java (hunk 7 above)
Comment (tvernum) [reply x-pack/plugin/security/src/main/java/org/elasticsearch/xpack/security/authz/RBACEngine.java:880]: I don't think this implementation can be considered thread safe. There's no memory barrier here, so there's no guarantee about the state of the `authorizedIndices` field when viewed from another thread.

The options are:
- simply state that it is not thread-safe and if callers require thread safety they need to synchronize
- mark `authorizedIndices` as volatile

File: x-pack/plugin/security/src/main/java/org/elasticsearch/xpack/security/authz/RBACEngine.java [hunk 11]
```diff
@@ -813,4 +854,99 @@ private static boolean isAsyncRelatedAction(String action) {
...
+
+        @Override
+        public boolean isEmpty() {
+            return getAuthorizedIndices().isEmpty();
+        }
+
+        @Override
+        public boolean contains(Object o) {
+            if (authorizedIndices == null) {
+                return predicate.test((String) o);
```
Comment (tvernum) [inline x-pack/plugin/security/src/main/java/org/elasticsearch/xpack/security/authz/RBACEngine.java:896]: Strictly speaking this violates the guarantees on the `Set` interface. You should return `false` if `o` is not a `String` (rather than throw CCE).
```suggestion
                if( o instanceof String s) {
                    return predicate.test(s);
                } else {
                    return false;
                }
```

File: x-pack/plugin/security/src/main/java/org/elasticsearch/xpack/security/authz/RBACEngine.java [hunk 12]
```diff
@@ -813,4 +854,99 @@ private static boolean isAsyncRelatedAction(String action) {
...
+        }
+
+        @Override
+        public boolean remove(Object o) {
+            throw new UnsupportedOperationException();
+        }
+
+        @Override
+        public boolean containsAll(Collection<?> c) {
+            return c.stream().allMatch(this::contains);
```
Comment (tvernum) [inline x-pack/plugin/security/src/main/java/org/elasticsearch/xpack/security/authz/RBACEngine.java:929]: Do we ever call this?
If so, then we probably want to delegate to `getAuthorizedIndices().containsAll` if the set has been populated:
```suggestion
             if (authorizedIndices == null) {
                 return c.stream().allMatch(this::contains);
             } else {
                 return getAuthorizedIndices().containsAll(c);
             }
```

**Review by tvernum — COMMENTED** [review]
There seems to be a lack of tests that this change actually achieves anything. The existing tests verify that you haven't broken any existing behaviour, but there's no test that would fail if someone changed the behaviour back so it was no longer lazy.

File: docs/changelog/81237.yaml (hunk 10 above)
Comment (ywangd) [reply docs/changelog/81237.yaml:2]: I changed it to:
> Avoid loading authorized indices when requested indices are all concrete names

File: x-pack/plugin/security/src/main/java/org/elasticsearch/xpack/security/authz/RBACEngine.java (hunk 7 above)
Comment (ywangd) [reply x-pack/plugin/security/src/main/java/org/elasticsearch/xpack/security/authz/RBACEngine.java:880]: You are right. It is indeed not thread safe. I changed the documentation to state that and warned about changing how this class is used.

File: x-pack/plugin/security/src/main/java/org/elasticsearch/xpack/security/authz/RBACEngine.java (hunk 11 above)
Comment (ywangd) [reply x-pack/plugin/security/src/main/java/org/elasticsearch/xpack/security/authz/RBACEngine.java:896]: I thought about it. But [JDK doc](https://docs.oracle.com/en/java/javase/17/docs/api/java.base/java/util/Set.html#contains(java.lang.Object)) says it can throw `ClassCastException `. So I didn't do anything. Also adding `instanceof` check can in theory hide subtle usage error. So I am not sure if it helps with anything.

File: x-pack/plugin/security/src/main/java/org/elasticsearch/xpack/security/authz/RBACEngine.java (hunk 12 above)
Comment (ywangd) [reply x-pack/plugin/security/src/main/java/org/elasticsearch/xpack/security/authz/RBACEngine.java:929]: No we don't call this method. But this is a good suggestion regardless. I updated.

**Comment by ywangd** [comment]
> There seems to be a lack of tests that this change actually achieves anything. The existing tests verify that you haven't broken any existing behaviour, but there's no test that would fail if someone changed the behaviour back so it was no longer lazy.

I added two assertions to existing tests to ensure that the returned `Set` from `resolveAuthorizedIndicesFromRole` is an instance of `AuthorizedIndicesSet`. Also added a new test to ensure the lazy behaviour of `AuthorizedIndicesSet`. 

PS: [Benchmark](https://github.com/elastic/elasticsearch/issues/79632#issuecomment-1021504869) showed that the change helped a great deal for requests with a single concrete name.

File: x-pack/plugin/security/src/test/java/org/elasticsearch/xpack/security/authz/RBACEngineTests.java [hunk 13]
```diff
@@ -1459,6 +1474,54 @@ public void testNoInfiniteRecursionForRBACAuthorizationInfoHashCode() {
...
+        doAnswer(invocation -> {
+            final String name = (String) invocation.getArguments()[0];
+            return authorizedNames.contains(name);
+        }).when(predicate).test(anyString());
+        final RBACEngine.AuthorizedIndicesSet authorizedIndicesSet = new RBACEngine.AuthorizedIndicesSet(supplier, predicate);
+
+        // Check with contains or containsAll do not trigger loading
+        final String name1 = randomFrom(allNames);
+        final String name2 = randomValueOtherThan(name1, () -> randomFrom(allNames));
+        final boolean containsAll = true; // randomBoolean();
```
Comment (tvernum) [inline x-pack/plugin/security/src/test/java/org/elasticsearch/xpack/security/authz/RBACEngineTests.java:1495]:
```suggestion
        final boolean containsAll = randomBoolean();
```

**Review by tvernum — APPROVED** [review]

File: x-pack/plugin/security/src/test/java/org/elasticsearch/xpack/security/authz/RBACEngineTests.java (hunk 13 above)
Comment (ywangd) [reply x-pack/plugin/security/src/test/java/org/elasticsearch/xpack/security/authz/RBACEngineTests.java:1495]: Err thanks!
